quickwit-search/src/search_client_pool.rs

```diff
@@ -247,8 +253,15 @@ impl SearchClientPool {
             job_order_key(left).cmp(&job_order_key(right))
         });
 
+        // Compute load (per node) threshold
+        let total_cost: Cost = jobs.iter().map(|v| v.cost()).sum();
+        let load_per_node_threshold =
+            ((total_cost as f64 / nodes.len() as f64) * (1.0 + LOAD_MARGIN)).ceil() as u64;
+        let load_per_node_threshold = load_per_node_threshold.max(MIN_LOAD_THRESHOLD);
+
         for job in jobs {
             sort_by_rendez_vous_hash(&mut nodes, job.split_id());
+
             // choose one of the the first two nodes based on least loaded
             let chosen_node_index: usize = if nodes.len() >= 2 {
```

Comment:
I think the whole point is to ditch this logic here.

The new algorithm will increase the chance of picking the node with the highest affinity.

Reply:
Hi, I would like to discuss a bit about the strategy here.

IMHO, there is no perfect load-balancer algorithm. I really think the idea of Francois by introducing cost threshold works, in terms of performance and simplicity.

Let me summarize a bit:
- Job is sorted desc by cost (same as previous strategy)
- The difference between old-new algorithm is applying a threshold = total_cost / num_node
- If any node reaches the threshold -> remove from selection

There are many cases of imbalance just like knapsack problem. So, I think we should focus on 2 popular cases:
- Cost is similar between split_id
- There are several super split_id(s) that each of them contains lot of documents, cost is high

In the first case, the algorithm works, just like old one. 
In the second case, if there are super split_id(s), most likely that there is only one node would pick 1 or 2 split_id, then reaches threshold, then be removed from next selection.

Additionally, new algorithm still respects `rendez_vous_hash` order for cache locality at leaf-node (IMHO, using `rendez_vous_hash` order is pretty good approach :+1: )

Reply:
I don't think you want to pick the node that is the least loaded out of the first two. This is useless. 

Thanks to Francois's solution you can pick the best one.

Reply:
In other words, `chosen_node_index = 0` all of the time.

Reply:
Hmm, I think code looks like:

```rust
// choose one of the the first two nodes based on least loaded
let chosen_node_index: usize = if nodes.len() >= 2 {
            if nodes[0].load > nodes[1].load {
                    1
                } else {
                    0
                }
            } else {
                0
            };
```
For each job -> node is sorted by `rendez_vous_hash`, not load. So the first two nodes (after sorted) does not mean lowest load.

Reply:
Why not simply take `nodes[0]`?

Reply:
I think take best one of two is good heuristic (take lower load from two). 

That's why I keep it. It's also harmless IMHO

Reply:
I don't think this is harmless at all.